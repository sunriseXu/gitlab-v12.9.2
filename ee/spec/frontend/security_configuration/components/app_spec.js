import { shallowMount } from '@vue/test-utils';

import { GlLink } from '@gitlab/ui';
import SecurityConfigurationApp from 'ee/security_configuration/components/app.vue';

describe('Security Configuration App', () => {
  let wrapper;
  const createComponent = (props = {}) => {
    wrapper = shallowMount(SecurityConfigurationApp, {
      propsData: {
        features: [],
        autoDevopsEnabled: false,
        latestPipelinePath: 'http://latestPipelinePath',
        autoDevopsHelpPagePath: 'http://autoDevopsHelpPagePath',
        helpPagePath: 'http://helpPagePath',
        pipelinesHelpPagePath: 'http://pipelineHelpPagePath',
        ...props,
      },
    });
  };

  afterEach(() => {
    wrapper.destroy();
  });

  const generateFeatures = n =>
    [...Array(n).keys()].map(i => ({
      name: `name-feature-${i}`,
      description: `description-feature-${i}`,
      link: `link-feature-${i}`,
    }));

  const getHelpLink = () => wrapper.find('header').find(GlLink);
  const getNotification = () => wrapper.find({ ref: 'callout' });
  const getPipelinesLink = () => getNotification().find('a');
  const getFeaturesTable = () => wrapper.find({ ref: 'featuresTable' });
  const getFeatureConfigStatus = () => wrapper.find({ ref: 'featureConfigStatus' });

  describe('header', () => {
    it('displays a link to the given help page', () => {
      const helpPagePath = 'http://foo';

      createComponent({ helpPagePath });

      expect(getHelpLink().attributes('href')).toBe(helpPagePath);
    });

    it.each`
      autoDevopsEnabled | latestPipelinePath         | expectedUrl
      ${true}           | ${'http://latestPipeline'} | ${'http://autoDevopsHelpPagePath'}
      ${false}          | ${'http://latestPipeline'} | ${'http://latestPipeline'}
      ${false}          | ${undefined}               | ${'http://pipelineHelpPagePath'}
    `(
      'displays a link to "$expectedUrl" when autoDevops is "$autoDevopsEnabled" and pipelinesPath is $latestPipelinePath',
      ({ autoDevopsEnabled, latestPipelinePath, expectedUrl }) => {
        createComponent({ autoDevopsEnabled, latestPipelinePath });

        expect(getPipelinesLink().attributes('href')).toBe(expectedUrl);
        expect(getPipelinesLink().attributes('rel')).toBe('noopener');
      },
    );
  });

  describe('features table', () => {
    it('displays a row for each given feature', () => {
      const features = generateFeatures(5);

      createComponent({ features });

      expect(wrapper.findAll({ ref: 'featureRow' }).length).toBe(5);
    });

    it('displays a given feature', () => {
      const features = generateFeatures(1);

      createComponent({ features });

      expect(getFeaturesTable().element).toMatchSnapshot();
    });

    it.each`
      configured | statusText
      ${true}    | ${'Configured'}
      ${false}   | ${'Not yet configured'}
    `(
      `displays "$statusText" if the given feature's configuration status is: "$configured"`,
      ({ configured, statusText }) => {
        const features = [{ configured }];

        createComponent({ features });

        expect(getFeatureConfigStatus().text()).toBe(statusText);
      },
    );
  });
});
