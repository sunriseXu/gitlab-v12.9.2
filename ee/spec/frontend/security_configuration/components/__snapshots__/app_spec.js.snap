// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Security Configuration App features table displays a given feature 1`] = `
<section
  class="mt-0"
>
  <div
    class="gl-responsive-table-row table-row-header text-2 font-weight-bold px-2 gl-text-gray-900"
    role="row"
  >
    <div
      class="table-section section-80"
    >
      
        Secure features
      
    </div>
     
    <div
      class="table-section section-20"
    >
      Status
    </div>
  </div>
   
  <div
    class="gl-responsive-table-row flex-md-column align-items-md-stretch px-2"
  >
    <div
      class="d-md-flex align-items-center"
    >
      <div
        class="table-section section-80 section-wrap pr-md-3"
      >
        <div
          class="table-mobile-header"
          role="rowheader"
        >
          
            Feature
          
        </div>
         
        <div
          class="table-mobile-content"
        >
          <div
            class="d-flex align-items-center justify-content-end justify-content-md-start"
          >
            <div
              class="text-2 gl-text-gray-900"
            >
              
                name-feature-0
              
            </div>
          </div>
           
          <div
            class="text-secondary"
          >
            
              description-feature-0
              
            <gl-link-stub
              aria-label="Feature documentation for name-feature-0"
              href="link-feature-0"
              target="_blank"
            >
              More information
            </gl-link-stub>
          </div>
        </div>
      </div>
       
      <div
        class="table-section section-20 section-wrap pr-md-3"
      >
        <div
          class="table-mobile-header"
          role="rowheader"
        >
          
            Status
          
        </div>
         
        <div
          class="table-mobile-content"
        >
          
            Not yet configured
          
        </div>
      </div>
    </div>
  </div>
</section>
`;
